File: Leetcode/Backtrace/17_letter-combinations-of-a-phone-number/main.py
from typing import List


class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return list()
        phone_map = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz",
        }
        combination = list()
        combinations = list()

        def backtrack(index: int):
            if index == len(digits):
                combinations.append("".join(combination))  # ['a', 'd', 'g'] → 'adg'
            else:
                num = digits[index]
                for letter in phone_map[num]:
                    combination.append(letter)
                    backtrack(index + 1)
                    combination.pop()

        backtrack(0)  # 执行内部的回溯子程序
        return combinations


if __name__ == '__main__':
    digits = "23"
    s = Solution()
    res = s.letterCombinations(digits)
    print(res)

File: Leetcode/Backtrace/17_letter-combinations-of-a-phone-number/main.cpp
// 17. 电话号码的字母组合
// Created by bob on 2022/3/8.
//
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include "KPrint.h"

using namespace std;

class Solution {
public:

    /**
     * 给定一个仅包含数字 2-9 的字符串，返回所有它能表示的字母组合。答案可以按 任意顺序 返回。
     * @param digits 电话号码
     * @return 它能表示的字母组合
     */
    vector<string> letterCombinations(string digits) {
        vector<string> combinations;
        if (digits.length() == 0) return combinations;
        unordered_map<char, string> phoneMap{
                {'2', "abc"},
                {'3', "def"},
                {'4', "ghi"},
                {'5', "jkl"},
                {'6', "mno"},
                {'7', "pqrs"},
                {'8', "tuv"},
                {'9', "wxyz"}
        };
        string combination;
        backtrack(phoneMap, digits, 0, combination, combinations);
        return combinations;
    }

    void backtrack(const unordered_map<char, string>& phoneMap, const string& digits,int index, string& combination, vector<string>& combinations) {
        if (index == digits.length()) {
            combinations.push_back(combination);
        } else {
            char num = digits[index];
            const string& letters = phoneMap.at(num);
            for (auto letter:letters) {
                combination.push_back(letter);
                backtrack(phoneMap, digits, index + 1, combination, combinations);
                combination.pop_back(); // ★临时的string容器需要清空
            }
        }
    }
};

int main() {
    Solution s = Solution();
    vector<string> vector = s.letterCombinations("23");
    printVector(vector);    // expect：ad, ae, af, bd, be, bf, cd, ce, cf
}